onnxruntime/contrib_ops/cuda/custom_layer_norm.cu

```diff
@@ -0,0 +1,307 @@
+https://github.com/microsoft/DeepSpeed/blob/4a3234e0ab3a6b0039741d7ea2fe0c1b82584763/csrc/transformer/normalize_kernels.cu.
+This layer requires template parameters since it relies fairly heavily on unrolling loops.
+The kernel is invoked when the n2(number of elements to reduce) is a multiple of a power of 2. 
+Default behavior for our purposes uses 256 threads for floats, and 128 threads for __half. 
+This restriction is a result of using the shift parameter to perform the minimum number of 
+register file shuffles necessary, which requires the number of threads in the secondary 
+reduction to be 1, 2, 4, 8, 16, or 32. The number of threads here corresponds to the number 
+of complete warps in the threadblock.
+For FP16, this kernel does not promote to FP32 in order to utilize the 2x throughput for
+__half2 instructions, and avoid the conversion overhead (1/8 of __hal2 arithmetic).
+For specific launch constraints, see the launch functions.
+*/
+
+template <int row_stride, int iterations>
+__global__ void custom_layer_norm(
+    float* vals,
+    const float* residual,
+    const float* gamma,
+    const float* beta,
+    float epsilon,
+    float* invvars = nullptr,
+    float* means = nullptr) {
+  constexpr int iteration_stride = row_stride / iterations;
+
+  cg::thread_block b = cg::this_thread_block();
+  cg::thread_block_tile<32> g = cg::tiled_partition<32>(b);
+
+  int row = blockIdx.x;
+  int id = threadIdx.x;
+  int gid = id / 32;
+
+  float vals_arr[iterations];
+  __shared__ float shr[iteration_stride >> 5];
+
+  float sum = 0.f;
+#pragma unroll
+  for (int i = 0; i < iterations; i++) {
+    vals_arr[i] = residual[row * row_stride + i * iteration_stride + id];
+    sum += vals_arr[i];
+  }
+
+  for (int stride = 1; stride < GPU_WARP_SIZE; stride *= 2) {
+    sum += WARP_SHFL_DOWN(sum, stride);
+  }
+
+  if (g.thread_rank() == 0) shr[gid] = sum;
+
+  b.sync();
+
+  if (g.thread_rank() < (iteration_stride >> 5)) sum = shr[g.thread_rank()];
+
+#if __CUDA_ARCH__ < 700
+  b.sync();
+#endif
+
+  for (int stride = 1; stride < (iteration_stride >> 5); stride *= 2) {
+    sum += WARP_SHFL_DOWN(sum, stride);
+  }
+
+  sum = WARP_SHFL(sum, 0);
+  float mean = sum / row_stride;
+  if (g.thread_rank() == 0) means[row] = mean;
+  float inv_variance, variance = 0.f;
+  for (int i = 0; i < iterations; i++) {
+    variance += (vals_arr[i] - mean) * (vals_arr[i] - mean);
+  }
+
+  for (int stride = 1; stride < GPU_WARP_SIZE; stride *= 2) {
+    variance += WARP_SHFL_DOWN(variance, stride);
+  }
+
+  if (g.thread_rank() == 0) shr[gid] = variance;
+
+  b.sync();
+
+  if (g.thread_rank() < (iteration_stride >> 5)) variance = shr[g.thread_rank()];
+
+  b.sync();
+
+  for (int stride = 1; stride < (iteration_stride >> 5); stride *= 2) {
+    variance += WARP_SHFL_DOWN(variance, stride);
+  }
+  variance = WARP_SHFL(variance, 0);
+  variance /= row_stride;
+  variance += epsilon;
+  inv_variance = rsqrtf(variance);
```

Comment:
would inv_stddev be a better name as it's 1/sqrt(variance)?

Reply:
similarly for invvars

Reply:
resolved offline, decided to keep the naming to be consistent with the existing kernel